Add version control dropdown to layout

drasmuss opened the pull request:
This is based on a template I made for NengoDL.  I made it a separate file so that it is easy for backends to change the template if they want (as it seems likely that different backends might want to tweak the display according to their own versioning system).

It is controlled by passing two flags to the sphinx build process, `building_version` and `releases`.  `building_version` specifies which version of the documentation is currently being built (with special name "latest" signifying the most up-to-date version).  `releases` is a semi-colon-separated list of past versions of the documentation we want to include (named "releases" because this probably corresponds to one documentation version for each public release).  It is assumed that these previous versions of the documentation have been built into subfolders of the root doc build, with the name of each subfolder corresponding to the name of the release.

For example
```
sphinx-build -b html docs docs/_build -A building_version=latest -A releases=v1.0.0;v1.1.0;v1.2.0
```

will give you a versioning dropdown that looks like

<img width="83" alt="capture" src="https://user-images.githubusercontent.com/1952220/45638359-5be4bf00-ba83-11e8-90b6-1cac9b6ef4ab.PNG">

You would build the other versions with, e.g.,
```
sphinx-build -b html docs docs/_build/v1.0.0 -A building_version=v1.0.0 -A releases=v1.0.0;v1.1.0;v1.2.0
```
(although I'd recommend setting up the doc build CI in such a way that those are saved automatically, rather than rebuilding them each time).

Note that if the `building_version` tag is omitted then there is no version dropdown displayed, so projects don't need to care about this unless they want to.

drasmuss committed:
Add version control dropdown to layout

This involved touching a bit of CSS, so I did some cleanup while
I was in there.

tbekolay commented:
Nice! Is this something that you can also specify through `conf.py`? Not the `building_version` so much, but I can see it being a bit easier to list the existing releases in `conf.py` and update that list on each release.

drasmuss commented:
Yep! You can add, e.g., `html_context = {"releases": "v1.0.0;v1.1.0"}` to your `conf.py` and specify it that way instead.

In my CI-based approach I just look for all subfolders starting with `v[0-9].*` and generate the `releases` list that way, so that I don't have to remember to update anything. But tracking it in `conf.py` would be a pretty minimal effort as well (and would let you use whatever names you want for releases).

tbekolay committed:
Validate building_version and releases

You have to opt in to this validation through adding
`nengo_sphinx_theme.ext.versions` to your extensions list.
More aspects of the versioning could be moved to this extension
in the future.

tbekolay committed:
Test version dropdown in feature test page

The deeply nested link was added because the most complicated
logic in the template is making sure that nested links go to
their correct locations in the other version of the docs.

This also fixes a small bug in the link to the main page when
on pages nested in a folder.

drasmuss commented on nengo_sphinx_theme/theme/versions.html:
```diff
@@ -8,20 +8,20 @@
             {% set prefix = pagename | list | select("equalto", "/") | map("replace", "/", "../") | join("") %}
             {% if building_version == "latest" %}
-                <div class="w-dropdown-link">latest</div>
+            <div class="w-dropdown-link">latest</div>
```
If we're taking a vote I prefer having the template well indented, as I spend a lot more time looking at that then the rendered HTML.  But I don't feel super strongly about it.

tbekolay replied:
Yeah, there's more work we should do on making the HTML look good anyway (mostly [whitespace control](http://jinja.pocoo.org/docs/2.10/templates/#whitespace-control)); was going to say in my review that we can drop this commit but you beat me to it ;)

drasmuss commented on nengo_sphinx_theme/theme/sidebar.html:
```diff
@@ -25,3 +25,6 @@
 
   </ul>
 </div>
+{% if building_version %}
```
I think we'll want this inside the div, so that it stays in place relative to the sidebar (with the sticky sidebar fix).  (Minor change I can do myself, I'm just writing it down here so I don't forget)

drasmuss replied:
Although I guess that would put it inside the outline box as well.  So maybe we just make this independently sticky? We'll see.

tbekolay replied:
I think you can set some limits to the stickiness, like having a top and bottom margin, so we might just have to set those such that the version dropdown is always still visible.

drasmuss replied:
Also `versions.html` has the `{% if building_version %}` check inside it already, so we can just include it here.

tbekolay replied:
It's technically faster not to include the template at all, but we're talking one or two milliseconds probably so whoever touches the file next can take it out

drasmuss commented:
Pushed a couple changes:

Some minor fixups (fbddc3a, 8e59f27)

I partially reverted 518b96f.  I played around a bit with trying to get the version dropdown to play nicely with the rest of the sidebar when scrolling, but I think it'd require more extensive CSS rewriting.  I figured that was more effort than we wanted to put in at this stage, so I just put it back in the main layout body.  The CSS change was necessary to preserve the width of the sidebar as the affix settings change (I rebased this branch onto the sticky sidebar fix branch, which revealed that issue).

Reverted c123020 as discussed.

tbekolay reviewed (approved):
Rebased this to master and fixed up the history, it's looking good to me! To be safe, could you try this branch out on NengoDL @drasmuss ? If it's working well then I'll merge this and make a `nengo-sphinx-theme` release so we can start switching over projects to the theme.

drasmuss commented:
Everything seems to be working well!
